roomserver/storage/postgres/user_room_keys_table.go

```diff
@@ -102,27 +102,31 @@ func (s *userRoomKeysStatements) SelectUserRoomPrivateKey(
 	return result, err
 }
 
-func (s *userRoomKeysStatements) BulkSelectUserNIDs(
-	ctx context.Context,
-	txn *sql.Tx,
-	senderKeys [][]byte,
-) (map[string]types.EventStateKeyNID, error) {
+func (s *userRoomKeysStatements) BulkSelectUserNIDs(ctx context.Context, txn *sql.Tx, senderKeys map[types.RoomNID][]ed25519.PublicKey) (map[string]types.UserRoomKeyPair, error) {
 	stmt := sqlutil.TxStmtContext(ctx, txn, s.selectUserNIDsStmt)
 
-	rows, err := stmt.QueryContext(ctx, pq.Array(senderKeys))
+	roomNIDs := make([]types.RoomNID, 0, len(senderKeys))
+	var senders [][]byte
+	for roomNID := range senderKeys {
+		roomNIDs = append(roomNIDs, roomNID)
+		for _, key := range senderKeys[roomNID] {
+			senders = append(senders, key)
+		}
+	}
+	rows, err := stmt.QueryContext(ctx, pq.Array(senders))
```

Comment:
This isn't what I was expecting, as this code will give incorrect results in some cases. This code just queries on the set of sender keys, ignoring the room NID for each key. This means it's possible to query this function asking for the user NID in room A with sender key X, and get back the user NID for room B with sender key X.

We either need to `WHERE room_nid` to filter DB-side or we need to only add `UserRoomKeyPair` to `result` if the room NID was asked for in `senderKeys`.